### PR TITLE
Fix #2556, make JUnit4 work in dotty compiled classes.

### LINKED ISSUE #2556: JUnit4 does not work with default methods
see https://github.com/scala/scala/pull/5369/commits/50b71b6c829faabfaa46197572fc4ddd03b7e9c1

### PR DESCRIPTION
**Pull request by DarkDimius**
https://github.com/scala/scala/commit/50b71b6c829faabfaa46197572fc4ddd03b7e9c1

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/MixinOps.scala [hunk 1]
```diff
@@ -59,10 +64,14 @@ class MixinOps(cls: ClassSymbol, thisTransform: DenotTransformer)(implicit ctx:
     def needsDisambiguation = competingMethods.exists(x=> !(x is Deferred)) // multiple implementations are available
     def hasNonInterfaceDefinition = competingMethods.exists(!_.owner.is(Trait)) // there is a definition originating from class
     meth.is(Method, butNot = PrivateOrAccessorOrDeferred) &&
-    (meth.owner.is(Scala2x) || needsDisambiguation || hasNonInterfaceDefinition ) &&
+    (meth.owner.is(Scala2x) || needsDisambiguation || hasNonInterfaceDefinition || needsJUnit4Fix(meth) ) &&
     isCurrent(meth)
   }
 
+  private def needsJUnit4Fix(meth: Symbol): Boolean = {
+    meth.annotations.nonEmpty && JUnit4Annotations.exists(annot => annot.exists && meth.hasAnnotation(annot))
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/transform/MixinOps.scala]: Why `annot.exists` if they are already filtered out in [`filter(_.exists)`](https://github.com/lampepfl/dotty/pull/2557/files#diff-18af2904c028f762167dfeae745bb6ceR19)?

**Comment by nicolasstucki** [comment]
Otherwise LGTM

File: tests/run/junitForwarders/C_1.scala [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+class C extends T
+
+object Test extends App {
+  def check(c: Class[_], e: String) = {
+    val s = c.getDeclaredMethods.sortBy(_.getName).map(m => s"${m.getName} - ${m.getDeclaredAnnotations.mkString(", ")}").mkString(";")
+    assert(s == e, s"found: $s\nexpected: $e")
+  }
+  check(classOf[C], "foo - @org.junit.Test()")
+  // TODO scala-dev#213: should `foo$` really carry the @Test annotation?
+  check(classOf[T], "$init$ - ;foo - @org.junit.Test();foo$ - @org.junit.Test()")
```
Comment (nicolasstucki) [inline tests/run/junitForwarders/C_1.scala]: `foo$ - @org.junit.Test()` is not there when running in vulpix.

Comment (DarkDimius) [inline tests/run/junitForwarders/C_1.scala]: It shouldn't be there:-)

**Pull request merged**